### PR TITLE
Sniffer asynchronous support

### PR DESCRIPTION
**Pull request by dgarske**
# Description

Sniffer asynchronous support.
* Adds stateful handling of DH shared secret computation in `SetupKeys`.
* Improved the decrypt handling to use internal functions and avoid generating alerts on failures.
* Fix for sniffer resume due to missing `sessionIDSz` broken in #4807.
* Fix sniffer test cases to split resume (session_ticket) tests.
* Add `snifftest` list of build features so test script can gate running resume test.

# Testing

`./configure --enable-sniffer --enable-asynccrypt CFLAGS="-DWOLFSSL_SNIFFER_CHAIN_INPUT" --enable-curve25519 --enable-curve448 && make check`

Also tested with Intel QuickAssist 8970 and QAT driver `QAT.L.4.16.0-00017.tar.gz`.

# Checklist

 - [X] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by anhu** [comment]
I feel like you might need a README.md.  I'm not really sure what any of this does.

File: src/sniffer.c [hunk 1]
```diff
@@ -6306,5 +6445,56 @@ int ssl_SetKeyCallback(SSLKeyCb cb, void* cbCtx)
...
+    while (srv) {
+        int nMax = maxEvents - eventCount, nReady = 0;
+        if (nMax <= 0) {
+            break; /* out of room in events list */
+        }
+        ret = wolfSSL_CTX_AsyncPoll(srv->ctx, events + nReady, nMax, flags, &nReady);
+        if (ret == 0) {
+            eventCount += nReady;
+
+            /* TODO: iterate event list and populate session userCtx into events[i]->context
```
Comment (anhu) [inline src/sniffer.c:6478]: Future work?  Do we have a place create issues to track TODOs?

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 2]
```diff
@@ -629,45 +709,69 @@ int main(int argc, char** argv)
...
+            chain = (void*)packet;
+            chainSz = header.caplen;
+            (void)isChain;
 #endif
 
-#if defined(WOLFSSL_SNIFFER_CHAIN_INPUT) && \
-    defined(WOLFSSL_SNIFFER_STORE_DATA_CB)
+#ifdef WOLFSSL_ASYNC_CRYPT
+            do {
+                WOLF_EVENT* events[1];
```
Comment (anhu) [inline sslSniffer/sslSnifferTest/snifftest.c:735]: I don't understand why its array of 1 element of pointer.

**Review by anhu — CHANGES_REQUESTED** [review]
Wow!! Big commit! Nice work.  I'll go through it again because I don't think I fully understand what this code does.

File: src/sniffer.c [hunk 3]
```diff
@@ -122,6 +126,7 @@
     #endif
 #endif
 
+#define ERROR_OUT(err, eLabel) { ret = (err); goto eLabel; }
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/sniffer.c:129]: This is a convention we use elsewhere for async.

File: src/sniffer.c [hunk 4]
```diff
@@ -2155,701 +2189,845 @@ static void ShowTlsSecrets(SnifferSession* session)
...
-
-        /* try and load static ephemeral */
-    #ifdef WOLFSSL_STATIC_EPHEMERAL
-        #ifndef SINGLE_THREADED
-        int keyLocked = 0;
-        if (ctx->staticKELockInit &&
-            wc_LockMutex(&ctx->staticKELock) == 0)
+        /* Static Curve25519 Key */
+        if (useCurveId == ECC_X25519) {
+            /* try and load static ephemeral */
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/sniffer.c:2664]: This term `static ephemeral` is awkward and was something we came up with for TLS v1.3 sniffing.... a fixed ephemeral. That is only useful for testing. In real world customers should use our wolfKeyMgr and not re-use ephemeral keys.

File: src/sniffer.c [hunk 5]
```diff
@@ -2155,701 +2189,845 @@ static void ShowTlsSecrets(SnifferSession* session)
...
-                        session->sslServer->arrays->preMasterSecret,
-                        &session->sslServer->arrays->preMasterSz,
-                        privKey, privKeySz,
-                        input, *sslBytes);
-                    PRIVATE_KEY_LOCK();
+                #endif
                 }
-            } while (ret == WC_PENDING_E);
         }
+        #if defined(ECC_TIMING_RESISTANT) && (!defined(HAVE_FIPS) || \
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/sniffer.c:2605]: Existing code that was moved. The `wc_ecc_set_rng` is not available in some FIPS versions and only used with `ECC_TIMING_RESISTANT`

File: src/sniffer.c [hunk 6]
```diff
@@ -2155,701 +2189,845 @@ static void ShowTlsSecrets(SnifferSession* session)
...
-        ret = wc_InitRsaKey_ex(&key, NULL, devId);
-        if (ret == 0) {
-            keyInit = 1;
-            keyBuf = ssl->buffers.key;
+#ifdef WOLFSSL_ASYNC_CRYPT
+    ret = wolfSSL_AsyncPop(ssl, &ssl->options.asyncState);
+    if (ret != WC_NOT_PENDING_E) {
+        /* Check for error */
+        if (ret < 0)
+            goto exit_sk;
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/sniffer.c:2331]: Following existing async coding flows.

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 2 above)
Comment (dgarske) [reply sslSniffer/sslSnifferTest/snifftest.c:735]: Same. Part of my todo for setting up a more advanced use case.

File: scripts/sniffer-tls13-gen.sh [hunk 7]
```diff
@@ -66,5 +66,97 @@ if [ "$1" == "hrr" ]; then
...
+    run_sequence $1
+    kill $tcpdump_pid
+}
+
+run_capture "ecc"           ""
+run_capture "ecc-resume"    "--enable-session-ticket"
+run_capture "dh"            "--disable-ecc"
+run_capture "dh-resume"     "--disable-ecc --enable-session-ticket"
+run_capture "x25519"        "--enable-curve25519 --disable-dh --disable-ecc"
+run_capture "x25519-resume" "--enable-curve25519 --disable-dh --disable-ecc --enable-session-ticket"
```
Comment (anhu) [inline scripts/sniffer-tls13-gen.sh:161]: Ed25519?

File: src/sniffer.c [hunk 8]
```diff
@@ -2155,701 +2189,861 @@ static void ShowTlsSecrets(SnifferSession* session)
...
+    int type;
+    union {
+    #ifndef NO_RSA
+        RsaKey rsa;
+    #endif
+    #if !defined(NO_DH) && defined(WOLFSSL_DH_EXTRA)
+        struct {
+            word32 pLen; /* modulus length */
+            word32 privKeySz;
+            byte   privKey[52]; /* max for TLS */
```
Comment (anhu) [inline src/sniffer.c:2202]: Should there be a `#define MAX_DH_PRIV_SZ 52` somewhere in a header?

File: src/sniffer.c [hunk 9]
```diff
@@ -2155,701 +2189,861 @@ static void ShowTlsSecrets(SnifferSession* session)
...
+    #ifdef HAVE_ECC
+        ecc_key ecc;
+    #endif
+    #ifdef HAVE_CURVE25519
+        curve25519_key x25519;
+    #endif
+    #ifdef HAVE_CURVE448
+        curve448_key x448;
+    #endif
+    } pub;
```
Comment (anhu) [inline src/sniffer.c:2227]: How come no DH and no RSA? 

Also, if ecc, x25519 and x448 are disabled at configure time, then you have an empty union.  Is that valid?

File: src/sniffer.c [hunk 10]
```diff
@@ -2155,701 +2189,861 @@ static void ShowTlsSecrets(SnifferSession* session)
...
 
-    #if defined(WOLFSSL_STATIC_EPHEMERAL) && !defined(SINGLE_THREADED)
-        if (keyLocked) {
-            wc_UnLockMutex(&ctx->staticKELock);
-        }
-    #endif
+    #ifdef HAVE_ECC
+        /* Static ECC Key */
+        if (useCurveId >= 0 && args->keyBuf == NULL
+        #ifdef HAVE_CURVE25519
```
Comment (anhu) [inline src/sniffer.c:2569]: Does defining `HAVE_CURVE25519` require `HAVE_ECC`?  Should this be outside the `HAVE_ECC` guard?

File: src/sniffer.c [hunk 11]
```diff
@@ -2155,701 +2189,861 @@ static void ShowTlsSecrets(SnifferSession* session)
...
 
-        /* left-padded with zeros up to the size of the prime */
-        if (ret == 0 && p_len > session->sslServer->arrays->preMasterSz) {
-            word32 diff = p_len - session->sslServer->arrays->preMasterSz;
-            XMEMMOVE(session->sslServer->arrays->preMasterSecret + diff,
-                    session->sslServer->arrays->preMasterSecret,
-                    session->sslServer->arrays->preMasterSz);
-            XMEMSET(session->sslServer->arrays->preMasterSecret, 0, diff);
-            session->sslServer->arrays->preMasterSz = p_len;
+                /* if curve not provided in key share data, then use private key curve */
```
Comment (anhu) [inline src/sniffer.c:2650]: If curve was provided in key share, should we do a consistency check against the private key curve?  (or is that done somewhere else?)

**Review by anhu — CHANGES_REQUESTED** [review]
Wow, this is great! Just a few more comments.  Also, I'd love to see a README.md so I can figure out how to take this stuff for a spin.

File: scripts/sniffer-testsuite.test [hunk 12]
```diff
@@ -14,53 +14,89 @@ fi
...
+        ./sslSniffer/sslSnifferTest/snifftest ./scripts/sniffer-tls13-dh-resume.pcap ./certs/statickeys/dh-ffdhe2048.pem 127.0.0.1 11111
+
+        RESULT=$?
+        [ $RESULT -ne 0 ] && echo -e "\nsnifftest TLS v1.3 DH failed\n" && exit 1
+    fi
+
+    # TLS v1.3 sniffer test X25519
+    if test $RESULT -eq 0
+    then
+        ./sslSniffer/sslSnifferTest/snifftest ./scripts/sniffer-tls13-x25519-resume.pcap ./certs/statickeys/x25519.pem 127.0.0.1 11111
```
[the remark this answers is not in the document]
Comment (dgarske) [reply scripts/sniffer-testsuite.test:80]: This is curve25519 and TLS denotes it using x25519. See https://datatracker.ietf.org/doc/html/rfc8446 see 4.2.7 `x25519(0x001D)`

File: src/sniffer.c (hunk 11 above)
Comment (dgarske) [reply src/sniffer.c:2650]: It is done in `wc_ecc_import_x963_ex` below.

**Review by SparkiDev — APPROVED** [review]

**Comment by anhu** [comment]
The following fails: `./configure --enable-sniffer --disable-rsa && make check`

```
PASS: scripts/openssl.test
PASS: scripts/external.test
FAIL: scripts/sniffer-testsuite.test
FAIL: testsuite/testsuite.test
PASS: scripts/resume.test
PASS: scripts/google.test
PASS: scripts/tls13.test
PASS: scripts/unit.test
```

**Review by anhu — CHANGES_REQUESTED** [review]
Due to test failure....

**Review by anhu — APPROVED** [review]
Excellent!!

**Pull request merged**